File: java/base/src/main/vsdk/toolkit/environment/light/Light.java
package vsdk.toolkit.environment.light;

import vsdk.toolkit.common.Entity;
import vsdk.toolkit.common.linealAlgebra.Vector3Dd;
import vsdk.toolkit.common.color.ColorRgb;
import vsdk.toolkit.environment.geometry.element.Ray;

public abstract class Light extends Entity
{
    private Vector3Dd position;
    private ColorRgb emission;
    private int id;

    /// This string should be used for specific application defined
    /// functionality. Can be null.
    private String name;

    protected Light(Vector3Dd position, ColorRgb emission)
    {
        this.position = position;
        this.emission = emission;
        this.id = 0;
        this.name = "";
    }

    public String getName()
    {
        return name;
    }

    public void setName(String n)
    {
        name = n;
    }

    public int getId()
    {
        return id;
    }

    public void setId(int i)
    {
        id = i;
    }

    public Vector3Dd getPosition()
    {
        return position;
    }

    public void setPosition(Vector3Dd position)
    {
        this.position = position;
    }

    public ColorRgb getEmission()
    {
        return emission;
    }

    public void setEmission(ColorRgb emission)
    {
        this.emission = emission;
    }

    /**
    True only for AmbientLight: shaders must add its contribution directly
    from getEmission() and skip shadow sampling entirely.
    @return whether this light is an ambient light
    */
    public boolean isAmbient()
    {
        return false;
    }

    /**
    Surface point -> light direction (normalized) and the shadow ray
    distance limit past which occluders no longer count (a large sentinel
    for lights with no finite position, e.g. DirectionalLight).
    @param surfacePoint point being shaded
    @return direction toward the light and the shadow ray distance limit
    */
    public abstract LightDirection getDirectionAndDistance(Vector3Dd surfacePoint);

    /**
    Attenuation contract carried over from the povCpp light hierarchy:
    lightSourceRay direction points from the surface toward the light,
    normalized.
    @param lightSourceRay surface-to-light ray
    @return attenuation factor in [0, 1] to multiply into the light's
    contribution
    */
    public abstract double evaluateLightResponseFactor(Ray lightSourceRay);

    public abstract Light copy();

    public record LightDirection(Vector3Dd direction, double maxShadowDistance)
    {
    }
}

File: typescript/base/src/main/vsdk/toolkit/environment/light/Light.ts
import { FundamentalEntity } from "../../common/FundamentalEntity.js";
import { Vector3Dd } from "../../common/linealAlgebra/Vector3Dd.js";
import { ColorRgb } from "../../common/color/ColorRgb.js";
import type { Ray } from "../geometry/element/Ray.js";
export abstract class Light extends FundamentalEntity {
    private name = "";
    private id = 0;
    public constructor(
        private position: Vector3Dd,
        private emission: ColorRgb,
    ) {
        super();
    }
    public getName(): string {
        return this.name;
    }
    public setName(name: string): void {
        this.name = name;
    }
    public getId(): number {
        return this.id;
    }
    public setId(id: number): void {
        this.id = id;
    }
    public getPosition(): Vector3Dd {
        return this.position;
    }
    public setPosition(position: Vector3Dd): void {
        this.position = position;
    }
    public getEmission(): ColorRgb {
        return this.emission;
    }
    public setEmission(emission: ColorRgb): void {
        this.emission = emission;
    }
    public isAmbient(): boolean {
        return false;
    }
    public abstract getDirectionAndDistance(surfacePoint: Vector3Dd): Light.LightDirection;
    public abstract evaluateLightResponseFactor(lightSourceRay: Ray): number;
    public abstract copy(): Light;
}
export namespace Light {
    export class LightDirection {
        public constructor(
            private readonly _direction: Vector3Dd,
            private readonly _maxShadowDistance: number,
        ) {}
        public direction(): Vector3Dd {
            return this._direction;
        }
        public maxShadowDistance(): number {
            return this._maxShadowDistance;
        }
    }
}
